refactor, test: Postion 포장, move interface화

dongKos opened the pull request:
Position을 불변 객체로 생성, Movable interface를 주입받아서 움직일 수 있도록 변경
Position, Car 단위 테스트 추가

dongKos committed:
refactor: positon 일급 컬렉션 포장

dongKos committed:
refactor, test: Postion 포장, move interface화

Position을 불변 객체로 생성, Movable interface를 주입받아서 움직일 수 있도록 변경
Position, Car 단위 테스트 추가

chs1234 commented on src/main/java/com/dongko/service/Movable.java:
```diff
@@ -0,0 +1,6 @@
+package com.dongko.service;
+
+@FunctionalInterface
```
`@FunctionalInterface` 어노테이션의 의미는 뭔가요ㅠㅠ
추천해주실 만한 아티클 있으면 추천 부탁드립니다.

dongKos replied:
`@FunctionalInterface`는 이 인터페이스는 함수가 하나만 있어야 한다고 강제하는 것입니다. 람다를 쓰려면 함수가 하나여야 해서 저런거 써주는거에요

저는 남의 코드 보다가 모르는게 나오면 chatGpt한테 물어보는 편인데 생각보다 아주 자세하게 설명 해줍니다.

chs1234 commented on src/main/java/com/dongko/domain/model/Position.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
@chs1234 

Equals와 HashCode 왜 재정의 해야하는지 알아보기..!
https://tecoble.techcourse.co.kr/post/2020-07-29-equals-and-hashCode/

cc. @wapago

dongKos replied:
굳굳 진짜 개중요하다고 생각합니다

dongKos committed:
refactor, test: Cars 일급 컬렉션 포장, Participants 리팩토링, 테스트코드 추가

Cars를 일급 컬렉션으로 포장함. Participants는 cars에게 메세지를 던진다.
단위테스트를 용이하게 하기 위해 Cars에 생성자를 추가함

chs1234 commented on src/main/java/com/dongko/domain/exception/InvalidPositionException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class InvalidPositionException extends RacingCarDomainException{
+    public InvalidPositionException() {
+        super("Position 값은 음수일 수 없습니다.");
```
RuntimeException
|
RacingCarDomainException
|                    
InvalidPositionException, InvalidCarNameException

다시 한번 확인하고 갑니다!

chs1234 commented on src/main/java/com/dongko/service/RandomThresholdMovingStrategy.java:
```diff
@@ -0,0 +1,10 @@
+
+public class RandomThresholdMovingStrategy implements Movable{
+
+    private static final int THRESHOLD = 4;
```
인터페이스는 진짜 개념적으로 고개를 끄덕끄덕 해도 막상 사용하려고 하면 버벅대게 만드는 놈인 거 같아요.

`움직일수 있나?` 라는 행동만 정의해두고
'어떻게'에 대한 전략은 각 구현 클래스들이 정의하는 걸로 다시 한번 이해했습니다.
예를 들어 추후에 움직일 수 있는 조건이 "자동차 이름이 포르쉐인 것만 전진"이라는 조건으로 변경되면
별도의 전략 클래스를 추가해서 구현하면 되겠군요!

dongKos replied:
인터페이스는 많이 써보는 수 밖에 없습니다. 개인적으로 사용하는 이유는 변화에 유연하게 대응하기 위해서 라고 생각합니다. 예를들어

```
class A {
 void someFunction(B b) {
   // do something
   b.doSomeThing();
 }
}
```

이라고 할때 A가 B가 아니라 C를 사용해야하도록 요구 사항이 바꼈다고 생각해봅시다. 그렇다면 코드는 이렇게 수정 될거에요.
```
class A {
 void someFunction(C c) {
   // do something
   c.doSomeThing();
 }
}
```

하지만 애초에 A 클래스의 someFunction을 구체적인 B,C 같은 클래스 타입이 아니라 interface를 받도록 했다면 구현 클래스가 B -> C로 바껴도 A의 코드를 바꿀일이 없겠죠? 또한 B를 저렇게 사용하는 A같은 애들이 백만개 라면? 생각만 해도 끔찎하죠? 하지만 인터페이스로 잘 분리해놨다면

A를 사용하는 클라이언트 코드만 바꾸면 됩니다

```
A a = new A(new B());
a.someFunction
```

이해가 되셨을까요? 다음 과제때는 일부러라도 인터페이스를 활용해서 구현 해보시길 추천드립니다.

chs1234 replied:
` B를 저렇게 사용하는 A같은 애들이 백만개 라면? 생각만 해도 끔찎하죠?`

"끔찎"에서 이해 빡 되었습니다ㅋㅋㅋ
정말 백만개라면 후.... 숨이 턱 막히네요
`결합도를 낮추고 응집도를 높힌다.` 는 말이 딱 이럴 때 쓰이는 거라는 생각이 듭니다..!!

chs1234 replied:
찬호형도 이 코멘트 꼭 한번 읽어봐주면 좋겠음!
cc. @wapago

chs1234 commented on src/main/java/com/dongko/Game.java:
```diff
@@ -1,23 +1,33 @@
     }
 
+    private void getMovingStrategy() {
+        this.movingStrategy = new RandomThresholdMovingStrategy();
```
나중에 전략이 변경되면 여기서 갈아끼우는 걸로..!

dongKos committed:
Merge pull request #4 from jvmoverloads/refactor/Cars-Participant

refactor, test: Cars 일급 컬렉션 포장, Participants 리팩토링, 테스트코드 추가
